### PR TITLE
zotero_enhancement_maps_generator

### PR DESCRIPTION
**Pull request by mtrojan-ub**
- not added to cronjob yet
- using C++17 features for map iteration

### CODE REVIEW

File: cpp/zotero_enhancement_maps_generator.cc [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    const auto endpoint_url(Zeder::GetFullDumpEndpointPath(flavour));
+    const std::unordered_set<unsigned> &entries_to_download {}; // intentionally empty
+    const std::unordered_set<std::string> columns_to_download {};  // intentionally empty
+    const std::unordered_map<std::string, std::string> filter_regexps {}; // intentionally empty
+    std::unique_ptr<Zeder::FullDumpDownloader::Params> downloader_params(new Zeder::FullDumpDownloader::Params(endpoint_url,
+                                                                         entries_to_download, columns_to_download, filter_regexps));
+
+    auto downloader(Zeder::FullDumpDownloader::Factory(Zeder::FullDumpDownloader::Type::FULL_DUMP, std::move(downloader_params)));
+    if (not downloader->download(&downloaded_flavour_entries))
+        LOG_ERROR("couldn't download full dump for " + Zeder::FLAVOUR_TO_STRING_MAP.at(flavour));
```
Comment (ruschein) [inline cpp/zotero_enhancement_maps_generator.cc:70]: Please include the error message of the downloader in parentheses.

File: cpp/zotero_enhancement_maps_generator.cc [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    std::unordered_set<std::string> issns;
+
+    std::unordered_set<Config::JournalParams::IniKey> ini_keys {
+        Config::JournalParams::IniKey::ONLINE_ISSN,
+        Config::JournalParams::IniKey::PRINT_ISSN
+    };
+    for (const auto ini_key : ini_keys) {
+        const std::string issn(ZederInterop::GetJournalParamsIniValueFromZederEntry(entry, flavour,
+                               ini_key));
+        if (MiscUtil::IsPossibleISSN(issn))
```
Comment (ruschein) [inline cpp/zotero_enhancement_maps_generator.cc:86]: Maybe issue a warning w/ Zeder ID, if an entry is not an ISSN.

File: cpp/zotero_enhancement_maps_generator.cc [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+
+void GenerateIssnToLicenseMap(const std::string &zotero_enhancement_maps_directory,
+                              const std::unordered_map<Zeder::Flavour, Zeder::EntryCollection> &downloaded_entries)
+{
+    const std::string map_path(zotero_enhancement_maps_directory + "/ISSN_to_licence.map");
+    LOG_INFO("Generating " + map_path);
+
+    std::map<std::string, std::string> issn_to_license_map;
+    for (const auto & [flavour, downloaded_flavour_entries] : downloaded_entries) {
```
Comment (ruschein) [inline cpp/zotero_enhancement_maps_generator.cc:101]: No space after the ampersand, please.

File: cpp/zotero_enhancement_maps_generator.cc [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                                       Config::JournalParams::IniKey::NAME));
+                const auto issns(GetZederEntryISSNs(downloaded_flavour_entry, flavour));
+                for (const auto &issn : issns)
+                    issn_to_license_map[issn] = license + " # " + name;
+            }
+        }
+    }
+
+    std::string output;
+    for (const auto & [issn, license] : issn_to_license_map)
```
Comment (ruschein) [inline cpp/zotero_enhancement_maps_generator.cc:118]: Ditto.

File: cpp/zotero_enhancement_maps_generator.cc [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        output += issn + "=" + license + '\n';
+
+    FileUtil::WriteStringOrDie(map_path, output);
+}
+
+
+}
```
Comment (ruschein) [inline cpp/zotero_enhancement_maps_generator.cc:125]: Missing comment.

File: cpp/zotero_enhancement_maps_generator.cc (hunk 1 above)
Comment (mtrojan-ub) [reply cpp/zotero_enhancement_maps_generator.cc:70]: Zeder::FullDumpDownloader does not provide an error message (this is NOT from Downloader.h)

**Pull request merged**